File: plugins/togetherai-skills/skills/together-audio/scripts/stt_transcribe.py
#!/usr/bin/env python3
"""
Together AI speech-to-text examples with the Python v2 SDK.

Demonstrates:
- transcription
- translation
- diarization
- timestamps

Usage:
    python stt_transcribe.py transcribe audio.mp3
    python stt_transcribe.py translate foreign_audio.mp3 --target-language en
    python stt_transcribe.py diarize meeting.mp3 --min-speakers 2 --max-speakers 5
    python stt_transcribe.py timestamps audio.mp3 --granularity word

Requirements:
    uv pip install "together>=2.0.0"
    export TOGETHER_API_KEY=your_key
"""

from __future__ import annotations

import argparse
from pathlib import Path

from together import Together

client = Together()


def transcribe(
    audio_path: Path,
    model: str,
    language: str | None,
    prompt: str | None,
    temperature: float | None,
) -> str:
    """Run a basic transcription request."""
    with open(audio_path, "rb") as audio_file:
        payload: dict[str, object] = {
            "file": audio_file,
            "model": model,
            "response_format": "json",
        }
        if language:
            payload["language"] = language
        if prompt:
            payload["prompt"] = prompt
        if temperature is not None:
            payload["temperature"] = temperature

        response = client.audio.transcriptions.create(**payload)
    print(f"Transcription: {response.text}")
    return response.text


def translate(
    audio_path: Path,
    model: str,
    target_language: str | None,
    prompt: str | None,
    temperature: float | None,
) -> str:
    """Translate audio, defaulting to English if no target language is provided."""
    with open(audio_path, "rb") as audio_file:
        payload: dict[str, object] = {
            "file": audio_file,
            "model": model,
        }
        if target_language:
            payload["language"] = target_language
        if prompt:
            payload["prompt"] = prompt
        if temperature is not None:
            payload["temperature"] = temperature

        response = client.audio.translations.create(**payload)
    print(f"Translation: {response.text}")
    return response.text


def diarize(audio_path: Path, model: str, min_speakers: int, max_speakers: int) -> None:
    """Run diarization and print speaker segments."""
    with open(audio_path, "rb") as audio_file:
        response = client.audio.transcriptions.create(
            file=audio_file,
            model=model,
            response_format="verbose_json",
            diarize=True,
            min_speakers=min_speakers,
            max_speakers=max_speakers,
        )
    if not response.speaker_segments:
        print("No speaker segments returned.")
        return

    for segment in response.speaker_segments:
        print(f"[{segment.speaker_id}] {segment.start:.1f}s-{segment.end:.1f}s: {segment.text}")


def timestamps(audio_path: Path, model: str, granularity: str) -> None:
    """Print timestamped transcription results."""
    with open(audio_path, "rb") as audio_file:
        response = client.audio.transcriptions.create(
            file=audio_file,
            model=model,
            response_format="verbose_json",
            timestamp_granularities=granularity,
        )
    print(f"Text: {response.text}")
    print(f"Language: {response.language}")
    print(f"Duration: {response.duration}s")

    if granularity == "word" and response.words:
        for word in response.words:
            print(f"'{word.word}' [{word.start:.2f}s - {word.end:.2f}s]")
        return

    if response.segments:
        for segment in response.segments:
            print(f"[{segment.start:.2f}s - {segment.end:.2f}s] {segment.text}")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Together AI STT examples")
    parser.add_argument(
        "mode",
        choices=("transcribe", "translate", "diarize", "timestamps"),
        help="Workflow to run",
    )
    parser.add_argument("audio_file", help="Path to the input audio file")
    parser.add_argument(
        "--model",
        default="openai/whisper-large-v3",
        help="Speech-to-text model",
    )
    parser.add_argument(
        "--language",
        default=None,
        help="Source language for transcription or target language for translation when appropriate",
    )
    parser.add_argument(
        "--target-language",
        default=None,
        help="Optional target language for translations",
    )
    parser.add_argument(
        "--prompt",
        default=None,
        help="Optional prompt to bias decoding",
    )
    parser.add_argument(
        "--temperature",
        type=float,
        default=None,
        help="Optional sampling temperature between 0.0 and 1.0",
    )
    parser.add_argument(
        "--granularity",
        choices=("segment", "word"),
        default="word",
        help="Timestamp granularity",
    )
    parser.add_argument(
        "--min-speakers",
        type=int,
        default=1,
        help="Minimum expected speakers for diarization",
    )
    parser.add_argument(
        "--max-speakers",
        type=int,
        default=5,
        help="Maximum expected speakers for diarization",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    audio_path = Path(args.audio_file)

    if args.mode == "transcribe":
        transcribe(
            audio_path=audio_path,
            model=args.model,
            language=args.language,
            prompt=args.prompt,
            temperature=args.temperature,
        )
        return

    if args.mode == "translate":
        translate(
            audio_path=audio_path,
            model=args.model,
            target_language=args.target_language or args.language,
            prompt=args.prompt,
            temperature=args.temperature,
        )
        return

    if args.mode == "diarize":
        diarize(
            audio_path=audio_path,
            model=args.model,
            min_speakers=args.min_speakers,
            max_speakers=args.max_speakers,
        )
        return

    timestamps(
        audio_path=audio_path,
        model=args.model,
        granularity=args.granularity,
    )


if __name__ == "__main__":
    main()

File: plugins/togetherai-skills/skills/together-audio/scripts/stt_transcribe.ts
#!/usr/bin/env -S npx tsx
/**
 * Together AI speech-to-text examples with the TypeScript SDK.
 *
 * Demonstrates:
 * - transcription
 * - translation
 * - diarization
 * - timestamps
 *
 * Usage:
 *   npx tsx stt_transcribe.ts transcribe audio.mp3
 *   npx tsx stt_transcribe.ts translate foreign_audio.mp3 --target-language en
 *   npx tsx stt_transcribe.ts diarize meeting.mp3 --min-speakers 2 --max-speakers 5
 *   npx tsx stt_transcribe.ts timestamps audio.mp3 --granularity word
 *
 * Requirements:
 *   npm install together-ai
 *   export TOGETHER_API_KEY=your_key
 */

import Together from "together-ai";
import { readFileSync } from "fs";

const client = new Together();

type Mode = "transcribe" | "translate" | "diarize" | "timestamps";

interface ParsedArgs {
  mode: Mode;
  audioFile: string;
  model: string;
  language?: string;
  targetLanguage?: string;
  prompt?: string;
  temperature?: number;
  granularity: "segment" | "word";
  minSpeakers: number;
  maxSpeakers: number;
}

function parseArgs(): ParsedArgs {
  const raw = process.argv.slice(2);
  const get = (flag: string): string | undefined => {
    const index = raw.indexOf(flag);
    if (index === -1) return undefined;
    return raw[index + 1];
  };

  const mode = raw[0] as Mode | undefined;
  const audioFile = raw[1];

  if (!mode || !audioFile) {
    console.error(
      "Usage: npx tsx stt_transcribe.ts <transcribe|translate|diarize|timestamps> <audio_file> [options]",
    );
    process.exit(1);
  }

  if (!["transcribe", "translate", "diarize", "timestamps"].includes(mode)) {
    console.error("Mode must be one of: transcribe, translate, diarize, timestamps");
    process.exit(1);
  }

  const temperature = get("--temperature");
  return {
    mode,
    audioFile,
    model: get("--model") ?? "openai/whisper-large-v3",
    language: get("--language"),
    targetLanguage: get("--target-language"),
    prompt: get("--prompt"),
    temperature: temperature ? Number(temperature) : undefined,
    granularity: (get("--granularity") ?? "word") as "segment" | "word",
    minSpeakers: Number(get("--min-speakers") ?? "1"),
    maxSpeakers: Number(get("--max-speakers") ?? "5"),
  };
}

function loadAudioFile(filePath: string): File {
  const buffer = readFileSync(filePath);
  const ext = filePath.split(".").pop()?.toLowerCase() ?? "wav";
  const mimeMap: Record<string, string> = {
    wav: "audio/wav",
    mp3: "audio/mpeg",
    m4a: "audio/mp4",
    webm: "audio/webm",
    flac: "audio/flac",
  };
  return new File([buffer], filePath, { type: mimeMap[ext] ?? "audio/wav" });
}

async function transcribe(args: ParsedArgs): Promise<void> {
  const payload: Record<string, unknown> = {
    file: loadAudioFile(args.audioFile),
    model: args.model,
    response_format: "json",
  };
  if (args.language) payload.language = args.language;
  if (args.prompt) payload.prompt = args.prompt;
  if (typeof args.temperature === "number") payload.temperature = args.temperature;

  const response = await client.audio.transcriptions.create(payload);
  console.log(`Transcription: ${response.text}`);
}

async function translate(args: ParsedArgs): Promise<void> {
  const payload: Record<string, unknown> = {
    file: loadAudioFile(args.audioFile),
    model: args.model,
  };
  if (args.targetLanguage ?? args.language) {
    payload.language = args.targetLanguage ?? args.language;
  }
  if (args.prompt) payload.prompt = args.prompt;
  if (typeof args.temperature === "number") payload.temperature = args.temperature;

  const response = await client.audio.translations.create(payload);
  console.log(`Translation: ${response.text}`);
}

async function diarize(args: ParsedArgs): Promise<void> {
  const response: any = await client.audio.transcriptions.create({
    file: loadAudioFile(args.audioFile),
    model: args.model,
    response_format: "verbose_json",
    diarize: true,
    min_speakers: args.minSpeakers,
    max_speakers: args.maxSpeakers,
  });

  if (!response.speaker_segments?.length) {
    console.log("No speaker segments returned.");
    return;
  }

  for (const segment of response.speaker_segments) {
    console.log(
      `[${segment.speaker_id}] ${segment.start.toFixed(1)}s-${segment.end.toFixed(1)}s: ${segment.text}`,
    );
  }
}

async function timestamps(args: ParsedArgs): Promise<void> {
  const response: any = await client.audio.transcriptions.create({
    file: loadAudioFile(args.audioFile),
    model: args.model,
    response_format: "verbose_json",
    timestamp_granularities: args.granularity,
  });

  console.log(`Text: ${response.text}`);
  console.log(`Language: ${response.language}`);
  console.log(`Duration: ${response.duration}s`);

  if (args.granularity === "word" && response.words) {
    for (const word of response.words) {
      console.log(`'${word.word}' [${word.start.toFixed(2)}s - ${word.end.toFixed(2)}s]`);
    }
    return;
  }

  if (response.segments) {
    for (const segment of response.segments) {
      console.log(`[${segment.start.toFixed(2)}s - ${segment.end.toFixed(2)}s] ${segment.text}`);
    }
  }
}

async function main(): Promise<void> {
  const args = parseArgs();

  if (args.mode === "transcribe") {
    await transcribe(args);
    return;
  }

  if (args.mode === "translate") {
    await translate(args);
    return;
  }

  if (args.mode === "diarize") {
    await diarize(args);
    return;
  }

  await timestamps(args);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
